src/Http/WebUtilities/src/FileBufferingWriteStream.cs

```diff
@@ -0,0 +1,300 @@
+            }
+            else
+            {
+                // If the MemoryStream is incapable of accomodating the content to be written
+                // spool to disk.
+                EnsureFileStream();
+
+                // Spool memory content to disk and clear in memory buffers. We no longer need to hold on to it
+                PagedByteBuffer.CopyTo(FileStream, clearBuffers: true);
+
+                FileStream.Write(buffer, offset, count);
+            }
+        }
+
+        /// <inheritdoc />
+        public override async Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            ThrowArgumentException(buffer, offset, count);
+            ThrowIfDisposed();
+
+            if (_bufferLimit.HasValue && _bufferLimit - BufferedLength < count)
+            {
+                Dispose();
+                throw new IOException("Buffer limit exceeded.");
+            }
+
+            // Allow buffering in memory if we're below the memory threshold once the current buffer is written.
+            var allowMemoryBuffer = (_memoryThreshold - count) >= PagedByteBuffer.Length;
+            if (allowMemoryBuffer)
+            {
+                // Buffer content in the MemoryStream if it has capacity.
+                PagedByteBuffer.Add(buffer, offset, count);
+                Debug.Assert(PagedByteBuffer.Length <= _memoryThreshold);
+            }
+            else
+            {
+                // If the MemoryStream is incapable of accomodating the content to be written
+                // spool to disk.
+                EnsureFileStream();
+
+                // Spool memory content to disk and clear in memory buffers. We no longer need to hold on to it
+                await PagedByteBuffer.CopyToAsync(FileStream, clearBuffers: true, cancellationToken);
+
+                await FileStream.WriteAsync(buffer, offset, count, cancellationToken);
+            }
+        }
+
+        public override void Flush()
+        {
+            // Do nothing.
+        }
+
+        /// <inheritdoc />
+        public override void SetLength(long value) => throw new NotSupportedException();
+
+        /// <summary>
+        /// Copies buffered content to <paramref name="destination"/>.
+        /// </summary>
+        /// <param name="destination">The <see cref="Stream" /> to copy to.</param>
+        /// <param name="bufferSize">The size of the buffer.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken" />.</param>
+        /// <returns>A <see cref="Task" /> that represents the asynchronous copy operation.</returns>
+        /// <remarks>
+        /// Users of this API do not need to reset the <see cref="Position" /> of this instance, prior to copying content.
+        /// </remarks>
+        public override async Task CopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken)
```

Comment:
This doesn't quite fit the stream CopyTo contract.
https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.copytoasync?view=netframework-4.7.2
"Copying begins at the current position in the current stream."  CopyTo/Async is also a Read operation and you set CanRead=false.
 Since you have an explicit reference to the type already, why not add something more specific like Move/Drain/FlushBufferAsync(Stream)? It should also clear the contents so you can re-use the instance (e.g. incremental flush).